src/code/sys_flashrom.c

```diff
@@ -3,42 +3,242 @@
-
-#pragma GLOBAL_ASM("asm/non_matchings/code/sys_flashrom/func_80185C24.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/code/sys_flashrom/SysFlashrom_ThreadEntry.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/code/sys_flashrom/func_80185DDC.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/code/sys_flashrom/func_80185EC4.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/code/sys_flashrom/func_80185F04.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/code/sys_flashrom/func_80185F64.s")
+//#include "prevent_bss_reordering.h"
+
+OSMesgQueue sFlashromMesgQueue;
+OSMesg sFlashromMesg[1];
+s32 sFlashromFlashIsGood;
+s32 sFlashromVendor;
+
+STACK(sSysFlashromStack, 0x1000);
+StackEntry sSysFlashromStackInfo;
+OSThread sSysFlashromThread;
+FlashromRequest sFlashromRequest;
+OSMesg sSysFlashromMsgBuf[1];
+
+s32 SysFlashrom_FlashIsGood(void) {
+    return sFlashromFlashIsGood;
+}
+
+const char* SysFlashrom_GetVendorStr(void) {
+    switch (sFlashromVendor) {
+        case FLASH_VERSION_MX_PROTO_A:
+            return "PROTO A";
+        case FLASH_VERSION_MX_A:
+            return "A";
+        case FLASH_VERSION_MX_C:
+            return "C";
+        case FLASH_VERSION_MEI:
+            return "D(NEW)";
+        case FLASH_VERSION_MX_B_AND_D:
+            return "B or D";
+        default:
+            return "UNKNOWN";
+    }
+}
+
+s32 SysFlashrom_CheckFlashType(void) {
+    u32 flashType;
+    u32 flashVendor;
+
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    osFlashReadId(&flashType, &flashVendor);
+    sFlashromVendor = flashVendor;
+    if (flashType == 0x11118001) {
+        if ((flashVendor == FLASH_VERSION_MX_PROTO_A) || (flashVendor == FLASH_VERSION_MX_A) ||
+            (flashVendor == FLASH_VERSION_MX_C) || (flashVendor == FLASH_VERSION_MEI) ||
+            (flashVendor == FLASH_VERSION_MX_B_AND_D)) {
+            return 0;
+        } else {
+            return -1;
+        }
+    }
+    return -1;
+}
+
+s32 SysFlashrom_InitFlash(void) {
+    osCreateMesgQueue(&sFlashromMesgQueue, sFlashromMesg, ARRAY_COUNT(sFlashromMesg));
+    osFlashInit();
+    sFlashromFlashIsGood = true;
+    if (SysFlashrom_CheckFlashType()) {
+        sFlashromFlashIsGood = false;
+        return -1;
+    }
+    return 0;
+}
+
+s32 SysFlashrom_ReadData(void* addr, s32 pageNum, s32 pageCount) {
+    OSIoMesg msg;
+
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    osInvalDCache(addr, pageCount * 128);
+    osFlashReadArray(&msg, 0, pageNum, addr, pageCount, &sFlashromMesgQueue);
+    osRecvMesg(&sFlashromMesgQueue, NULL, OS_MESG_BLOCK);
+    return 0;
+}
+
+s32 SysFlashrom_ErasePage(u32 page) {
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    return osFlashSectorErase(page);
+}
+
+s32 SysFlashrom_WriteData(u8* addr, u32 pageNum, u32 pageCount) {
+    OSIoMesg msg;
+    s32 result;
+    u32 i;
+
+    if (!SysFlashrom_FlashIsGood()) {
+        return -1;
+    }
+    if (pageNum & 0x7F) {
```

Comment:
Maybe this expects a multiple of 0x80 / 128 ?
```suggestion
    if ((pageNum % 128) != 0) {
```